s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/RecordProcessor.java

```diff
@@ -53,44 +53,39 @@ public static List<SourceRecord> processRecords(final Iterator<List<AivenS3Sourc
+            if (aivenS3SourceRecord != null) {
+                final SourceRecord sourceRecord = createSourceRecord(aivenS3SourceRecord, s3SourceConfig, keyConverter,
+                        valueConverter, conversionConfig, outputWriter, failedObjectKeys, offsetManager);
+                results.add(sourceRecord);
+            }
         }
 
         LOGGER.info("Number of records sent {}", results.size());
         return results;
     }
 
-    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
-    static List<SourceRecord> createSourceRecords(final List<AivenS3SourceRecord> aivenS3SourceRecordList,
+    static SourceRecord createSourceRecord(final AivenS3SourceRecord aivenS3SourceRecord,
             final S3SourceConfig s3SourceConfig, final Optional<Converter> keyConverter, final Converter valueConverter,
             final Map<String, String> conversionConfig, final OutputWriter outputWriter,
             final Set<String> failedObjectKeys, final OffsetManager offsetManager) {
 
-        final List<SourceRecord> sourceRecordList = new ArrayList<>();
-        for (final AivenS3SourceRecord aivenS3SourceRecord : aivenS3SourceRecordList) {
-            LOGGER.info(" ******* CSR key ******** {}", aivenS3SourceRecord.getObjectKey());
-            final String topic = aivenS3SourceRecord.getToTopic();
-            final Optional<SchemaAndValue> keyData = keyConverter
-                    .map(c -> c.toConnectData(topic, aivenS3SourceRecord.key()));
-
-            outputWriter.configureValueConverter(conversionConfig, s3SourceConfig);
-            valueConverter.configure(conversionConfig, false);
-            try {
-                final SchemaAndValue schemaAndValue = valueConverter.toConnectData(topic, aivenS3SourceRecord.value());
-                offsetManager.updateCurrentOffsets(aivenS3SourceRecord.getPartitionMap(),
-                        aivenS3SourceRecord.getOffsetMap());
-                aivenS3SourceRecord.setOffsetMap(offsetManager.getOffsets().get(aivenS3SourceRecord.getPartitionMap()));
-                sourceRecordList.add(aivenS3SourceRecord.getSourceRecord(topic, keyData, schemaAndValue));
-            } catch (DataException e) {
-                LOGGER.error("Error in reading s3 object stream " + e.getMessage());
-                failedObjectKeys.add(aivenS3SourceRecord.getObjectKey());
-                throw e;
-            }
+        final String topic = aivenS3SourceRecord.getToTopic();
+        final Optional<SchemaAndValue> keyData = keyConverter
+                .map(c -> c.toConnectData(topic, aivenS3SourceRecord.key()));
+
+        outputWriter.configureValueConverter(conversionConfig, s3SourceConfig);
+        valueConverter.configure(conversionConfig, false);
+        try {
+            final SchemaAndValue schemaAndValue = valueConverter.toConnectData(topic, aivenS3SourceRecord.value());
+            offsetManager.updateCurrentOffsets(aivenS3SourceRecord.getPartitionMap(),
```

Comment:
Should update Current offsets be in a finally block?
It should be updated at the end of every processed record

Reply:
For every partition map there could be multiple object keys. So here, before sending it back to OffsetReader through SourceRecords, we update them and set them back into SourceRecord.

Reply:
Yea, if connectData throws an exception, we should not update offsets. So we are good here and updating only when no exceptions.

Reply:
Sorry I saw the responses and thought I had responded. I am happy to leave as-is, but my assumption is we wouldn't increment the offset until it was successfully completed, so on an exception,it would be the same location and be a zero-op but this is fine too.